api/src/main/java/org/apache/iceberg/util/StructProjection.java

```diff
@@ -189,16 +189,12 @@ public int size() {
 
   @Override
   public <T> T get(int pos, Class<T> javaClass) {
-    if (struct == null) {
-      // Return a null struct when projecting a nested required field from an optional struct.
-      // See more details in issue #2738.
-      return null;
-    }
-
     int structPos = positionMap[pos];
-
     if (nestedProjections[pos] != null) {
-      return javaClass.cast(nestedProjections[pos].wrap(struct.get(structPos, StructLike.class)));
+      StructLike nestedStruct = struct.get(structPos, StructLike.class);
+      return nestedStruct == null
```

Comment:
minor: ternary operators split on multiple lines are bit hard to read, in my view. I'd consider using `if`.